features/include/pcl/features/impl/gasd.hpp

```diff
@@ -0,0 +1,414 @@
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef PCL_FEATURES_IMPL_GASD_H_
+#define PCL_FEATURES_IMPL_GASD_H_
+
+#include <pcl/features/gasd.h>
+#include <pcl/common/transforms.h>
+#include <pcl/point_types_conversion.h>
+
+#include <vector>
+
+namespace pcl
+{
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT> void
+  GASDEstimation<PointInT, PointOutT>::compute (PointCloudOut &output)
+  {
+    if (!Feature<PointInT, PointOutT>::initCompute ())
+    {
+      output.width = output.height = 0;
+      output.points.clear ();
+      return;
+    }
+
+    // Resize the output dataset
+    output.width = output.height = 1;
+    output.points.resize (1);
+
+    // Copy header and is_dense flag from input
+    output.header = surface_->header;
+    output.is_dense = surface_->is_dense;
+
+    // Perform the actual feature computation
+    computeFeature (output);
+
+    Feature<PointInT, PointOutT>::deinitCompute ();
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT> void
+  GASDEstimation<PointInT, PointOutT>::computeAlignmentTransform ()
+  {
+    Eigen::Vector4f centroid;
+    Eigen::Matrix3f covariance_matrix;
+
+    // compute centroid of the object's partial view
+    pcl::compute3DCentroid (*surface_, *indices_, centroid);
+
+    // compute covariance matrix from points and centroid of the object's partial view
+    pcl::computeCovarianceMatrix (*surface_, *indices_, centroid, covariance_matrix);
+
+    Eigen::Matrix3f eigenvectors;
+    Eigen::Vector3f eigenvalues;
+
+    // compute eigenvalues and eigenvectors of the covariance matrix
+    pcl::eigen33 (covariance_matrix, eigenvectors, eigenvalues);
+
+    // z axis of the reference frame is the eigenvector associated with the minimal eigenvalue
+    Eigen::Vector3f z_axis = eigenvectors.col (0);
+
+    // if angle between z axis and viewing direction is in the [-90 deg, 90 deg] range, then z axis is negated
+    if (z_axis.dot (view_direction_) > 0)
+    {
+      z_axis = z_axis * -1;
```

Comment:
If you want to switch sign in something the best way to do it is simply 

z = -z;